File: android-native/app/src/main/java/flexr/social/app/data/repository/SafetyRepository.kt
package flexr.social.app.data.repository

import flexr.social.app.core.common.ServerTime
import flexr.social.app.core.network.apiCall
import flexr.social.app.data.remote.FlexrApi
import flexr.social.app.data.remote.dto.BlockRequestDto
import flexr.social.app.data.remote.dto.ReportRequestDto
import flexr.social.app.domain.model.BlockedUser
import flexr.social.app.domain.model.ModerationNotice
import flexr.social.app.domain.model.ReportAck
import javax.inject.Inject
import javax.inject.Singleton

/** Melden und Blockieren. Entspricht backend/app/routers/safety.py. */
@Singleton
class SafetyRepository @Inject constructor(
    private val api: FlexrApi,
) {

    /**
     * Meldet einen Nutzer. Die Antwort ist die Empfangsbestätigung mit
     * Aktenzeichen — sie muss dem Melder angezeigt werden (Art. 16 Abs. 4 DSA).
     */
    suspend fun report(userId: String, reason: String): ReportAck {
        val ack = apiCall { api.report(ReportRequestDto(userId, reason.trim())) }
        return ReportAck(reference = ack.reference, message = ack.message)
    }

    /** Laufende Beschränkung des eigenen Kontos, oder null wenn keine besteht. */
    suspend fun moderationNotice(): ModerationNotice? =
        apiCall { api.moderationNotice() }?.let { dto ->
            ModerationNotice(
                reason = dto.reason,
                mutedUntil = ServerTime.parse(dto.mutedUntil),
                appealHint = dto.appealHint,
            )
        }

    /** Blockieren wirkt beidseitig: das Match verschwindet auf beiden Seiten. */
    suspend fun block(userId: String) {
        apiCall { api.block(BlockRequestDto(userId)) }
    }

    suspend fun blockedUserIds(): List<String> = apiCall { api.listBlocks() }

    /** Blockierte Personen mit Name, Alter und Vorschaubild für die Verwaltungsliste. */
    suspend fun blockedUsers(): List<BlockedUser> =
        apiCall { api.listBlockedUsers() }.map { dto ->
            BlockedUser(
                userId = dto.userId,
                name = dto.name,
                age = dto.age,
                photoUrl = dto.photoUrl,
                blockedAt = ServerTime.parse(dto.blockedAt),
            )
        }

    suspend fun unblock(userId: String) {
        apiCall { api.unblock(userId) }
    }
}

File: ios/FLEXR/Data/Repository/SafetyRepository.swift
import Foundation

/// Melden und Blockieren. Entspricht backend/app/routers/safety.py.
@MainActor
final class SafetyRepository {

    private let api: FlexrAPI

    init(api: FlexrAPI) {
        self.api = api
    }

    /// Meldet einen Nutzer. Die Antwort ist die Empfangsbestätigung mit
    /// Aktenzeichen — sie muss dem Melder angezeigt werden (Art. 16 Abs. 4 DSA).
    func report(userID: String, reason: String) async throws -> ReportAck {
        let ack = try await api.report(
            ReportRequestDTO(
                reportedUserId: userID,
                reason: reason.trimmingCharacters(in: .whitespacesAndNewlines)
            )
        )
        return ReportAck(reference: ack.reference, message: ack.message)
    }

    /// Laufende Beschränkung des eigenen Kontos, oder nil wenn keine besteht.
    func moderationNotice() async throws -> ModerationNotice? {
        guard let dto = try await api.moderationNotice() else { return nil }
        return ModerationNotice(
            reason: dto.reason,
            mutedUntil: ServerTime.parse(dto.mutedUntil),
            appealHint: dto.appealHint
        )
    }

    /// Blockieren wirkt beidseitig: das Match verschwindet auf beiden Seiten.
    func block(userID: String) async throws {
        try await api.block(BlockRequestDTO(userId: userID))
    }

    func blockedUserIDs() async throws -> [String] {
        try await api.listBlocks()
    }

    func unblock(userID: String) async throws {
        try await api.unblock(userID: userID)
    }
}

/// Foto-Verifizierung (blauer Haken).
///
/// Der Server gibt die Anweisung vor ("Schau direkt in die Kamera"), das Selfie entsteht live über die
/// Kamera — kein Galerie-Upload. Das ist der Liveness-Schutz: nur eine echte
/// Person vor der Kamera kann die Aufnahme liefern.
@MainActor
final class VerificationRepository {

    private static let mimeType = "image/jpeg"

    private let api: FlexrAPI

    init(api: FlexrAPI) {
        self.api = api
    }

    func status() async throws -> VerificationState {
        try await api.verificationStatus().toDomain()
    }

    func start() async throws -> VerificationState {
        try await api.startVerification().toDomain()
    }

    /// Lädt die drei Aufnahmen hoch und reicht sie ein. Die Reihenfolge muss
    /// exakt der ausgegebenen Anweisungsliste entsprechen, sonst weist das Backend
    /// die Einreichung zurück.
    func submit(selfies: [(prompt: String, data: Data)]) async throws -> VerificationState {
        var uploaded: [VerificationSelfieDTO] = []
        for selfie in selfies {
            let presign = try await api.presignSelfie(
                PresignPhotoRequestDTO(contentType: Self.mimeType)
            )
            try await api.upload(
                to: presign.uploadUrl,
                contentType: Self.mimeType,
                data: selfie.data
            )
            uploaded.append(
                VerificationSelfieDTO(prompt: selfie.prompt, objectKey: presign.objectKey)
            )
        }
        return try await api.submitVerification(
            VerificationSubmitRequestDTO(selfies: uploaded)
        ).toDomain()
    }
}
